protocols/bgp/server/bmp_server.go

```diff
@@ -52,15 +53,116 @@ func NewServer(keepalivePeriod time.Duration) *BMPServer {
 	return b
 }
 
+// Listen starts a listener for routers to start the BMP connection
+// the listener needs to be closed by calling Close() on the BMPServer
+func (b *BMPServer) Listen(addr string) error {
+	tcp, err := net.ResolveTCPAddr("tcp", addr)
+	if err != nil {
+		log.WithError(err).WithFields(log.Fields{
+			"component": "bmp_server",
+			"address":   addr,
+		}).Info("Unable to resolve address")
+		return err
+	}
+	l, err := net.ListenTCP("tcp", tcp)
+	if err != nil {
+		log.WithError(err).WithFields(log.Fields{
+			"component": "bmp_server",
+			"address":   addr,
+		}).Infof("Unable to listen on %s", tcp.String())
+		return err
+	}
+	b.listener = l
+
+	for {
+		c, err := l.Accept()
+		if err != nil {
+			log.WithError(err).WithFields(log.Fields{
+				"component": "bmp_server",
+				"address":   addr,
+			}).Infof("Unable to accept on %s", tcp.String())
+			return err
+		}
+		go b.handleConnection(c)
+	}
+}
+
+func (b *BMPServer) handleConnection(c net.Conn) {
+	remoteIP := c.RemoteAddr().(*net.TCPAddr).IP
+
+	var r *Router
+	for _, router := range b.getRouters() {
+		if router.passive && remoteIP.Equal(router.address) {
+			r = router
+		}
+	}
+	if r == nil {
+		log.WithFields(log.Fields{
+			"component":      "bmp_server",
+			"remote_address": c.RemoteAddr(),
+		}).Error("dropping unconfigured connection")
+		c.Close()
+		return
+	}
+
+	if atomic.LoadUint32(&r.established) == 1 {
+		log.WithFields(log.Fields{
+			"component":      "bmp_server",
+			"remote_address": c.RemoteAddr(),
+			"router":         r.Name(),
+		}).Error("router already connected, dropping connection")
+		c.Close()
+		return
+	}
+
+	b.establishConnection(c, r, true)
+}
+
+func (b *BMPServer) establishConnection(c net.Conn, r *Router, passive bool) error {
```

Comment:
Not happy with this naming. Let's rename establishConnection to handleConnection. The current handleConnection should be validateConnection() bool. Move the call of establishConnection into the Listen() method and call it only if validateConnection() returned true.

Reply:
I changed the names and moved the call.